From f4f13305a0319ba14b6fc0ba2168b5169d8445b5 Mon Sep 17 00:00:00 2001
From: Stephane Desneux <stephane.desneux@iot.bzh>
Date: Tue, 16 Jul 2019 17:23:50 +0000
Subject: [PATCH] aglsetup: add -t|--topic and -V|--version options

New options are:
* -t|--topic   : add a 'topic' to current setup (default: empty)
* -V|--version : display aglsetup version

The topic value can be any string and it is propagated to aglsetup.manifest,
which in turn is used by distro-build-manifest to put it in DIST_BUILD_TOPIC
variable in all build manifests (deploy dir, target image, sdk)

Using the --topic option to specify a unique identifier will help to provide a
common ID for all builds in the same "family" even if features and layers differ.

For this purpose, a topic could be:
* a gerrit review id + a patchset number
* a project name (git repo) + a Change-Id
* a random UUID
* ...

When run with -V or --version, aglsetup writes its version on stdout and sets
the variable AGLSETUP_VERSION then exits. As aglsetup can only be called by
being sourced, the AGLSETUP_VERSION variable is finally set in the caller
environment.

The version can be used for various purposes. The immediate goal is to detect
if --topic option is supported or not.

Bug-AGL: SPEC-2646

Change-Id: I0b68dfa297509dac07e9b2942948631cfc13c319
Signed-off-by: Stephane Desneux <stephane.desneux@iot.bzh>
---
 scripts/.aglsetup_genconfig.bash     | 46 +++++++++++++++++++++++++---
 scripts/aglsetup.sh                  |  2 +-
 scripts/distro-manifest-generator.sh |  9 ++++--
 3 files changed, 48 insertions(+), 9 deletions(-)

diff --git a/scripts/.aglsetup_genconfig.bash b/scripts/.aglsetup_genconfig.bash
index 6e091e2f..5e80b1f7 100755
--- a/scripts/.aglsetup_genconfig.bash
+++ b/scripts/.aglsetup_genconfig.bash
@@ -4,7 +4,7 @@
 #
 # The MIT License (MIT)
 #
-# Copyright (c) 2016 Stéphane Desneux <sdx@iot.bzh>
+# Copyright (c) 2016-2019 Stéphane Desneux <sdx@iot.bzh>
 #           (c) 2016 Jan-Simon Möller <jsmoeller@linuxfoundation.org>
 #
 # Permission is hereby granted, free of charge, to any person obtaining a copy
@@ -31,11 +31,12 @@
 # turn execute (source) generated instructions back in the parent shell,
 # whether it's bash, zsh, or any other supported shell
 
-VERSION=1.1.0
+VERSION=1.2.0
 DEFAULT_MACHINE=qemux86-64
 DEFAULT_BUILDDIR=./build
 VERBOSE=0
-DEBUG=0
+SHOWVERSION=0
+DEBUG=${DEBUG:-0}
 
 #SCRIPT=$(basename $BASH_SOURCE)
 SCRIPT=aglsetup.sh
@@ -48,7 +49,7 @@ function error() { echo "ERROR: $@" >&2; return 1; }
 function verbose() { [[ $VERBOSE == 1 ]] && echo "$@" >&2; return 0; }
 function debug() { [[ $DEBUG == 1 ]] && echo "DEBUG: $@" >&2; return 0;}
 
-info "------------ $SCRIPT: Starting"
+debug "------------ $SCRIPT: starting with command line arguments: $@"
 
 #compute AGL_REPOSITORIES
 AGL_REPOSITORIES=$(for x in $(ls -d $METADIR/meta-*/templates/{machine,feature} $METADIR/bsp/*/templates/machine 2>/dev/null); do echo $(basename $(dirname $(dirname $x))); done | sort -u)
@@ -161,11 +162,22 @@ Options:
           'timestamp' : Use a generated time stamp (UTC).
           'value:<revision>' : Use <revision> explicitly.
           'none' : Do nothing.
+   -t|--topic <value>
+      Specify an optional topic for this setup.
+      If specified, the topic will be propagated in build manifests:
+         - in deployment dir: tmp/deploy/images/*/build-info
+         - in target image: /etc/platform-info/build
+         - in SDK: tmp/deploy/sdk/*.build-info
    -v|--verbose
       verbose mode
       default: false
+   -V|--version
+      display version, set AGLSETUP_VERSION variable with version value and exit
    -d|--debug
       debug mode
+      for early debug, set env variable DEBUG. 
+      for eaxmple:
+           DEBUG=1 source aglsetup.sh -V
       default: false
    -h|--help
       get some help
@@ -293,7 +305,7 @@ function find_feature_dependency() {
 
 GLOBAL_ARGS=( "$@" )
 debug "Parsing arguments: $@"
-TEMP=$(getopt -o m:b:r:s:fvdh --long machine:,builddir:,rpm-revision:,script:,force,verbose,debug,help -n $SCRIPT -- "$@")
+TEMP=$(getopt -o m:b:r:t:s:fvVdh --long machine:,builddir:,rpm-revision:,topic:,script:,force,verbose,version,debug,help -n $SCRIPT -- "$@")
 [[ $? != 0 ]] && { usage; exit 1; }
 eval set -- "$TEMP"
 
@@ -305,6 +317,7 @@ BUILDDIR=$DEFAULT_BUILDDIR
 SETUPSCRIPT=
 FORCE=
 RPMREVISION=
+TOPIC=
 SETUP_MANIFEST=aglsetup.manifest
 
 while true; do
@@ -314,7 +327,9 @@ while true; do
 		-s|--setupscript)  SETUPSCRIPT=$2; shift 2;;
 		-f|--force)        FORCE=1; shift;;
 		-r|--rpm-revision) RPMREVISION=$2; shift 2;;
+		-t|--topic)        TOPIC=$2; shift 2;;
 		-v|--verbose)      VERBOSE=1; shift;;
+		-V|--version)      SHOWVERSION=1; shift;;
 		-d|--debug)        VERBOSE=1; DEBUG=1; shift;;
 		-h|--help)         HELP=1; shift;;
 		--)                shift; break;;
@@ -324,6 +339,24 @@ done
 
 [[ "$HELP" == 1 ]] && { usage; exit 0; }
 
+if [[ "$SHOWVERSION" == 1 ]]; then
+	# display version on stdout
+	echo "$VERSION"
+
+	# generate output script if requested by caller
+	if [[ -n "$SETUPSCRIPT" ]]; then
+		cat <<EOF >$SETUPSCRIPT
+AGLSETUP_VERSION=$VERSION
+EOF
+	fi
+
+	# IMPORTANT: exit successfully
+	# older aglsetup scripts with version <1.2.0 will fail with option --version
+	exit 0
+fi
+
+info "------------ $SCRIPT: Starting"
+
 verbose "Command line arguments: ${GLOBAL_ARGS[@]}"
 
 # the remaining args are the features
@@ -510,6 +543,9 @@ DIST_METADIR="$METADIR"
 # timestamp
 DIST_SETUP_TS="$(date -u +%Y%m%d_%H%M%S_%Z)"
 
+# topic
+DIST_SETUP_TOPIC="$TOPIC"
+
 # ------------ end of $SCRIPT fragment --------
 EOF
 info "OK"
diff --git a/scripts/aglsetup.sh b/scripts/aglsetup.sh
index 21a1e807..3bb6e0cd 100644
--- a/scripts/aglsetup.sh
+++ b/scripts/aglsetup.sh
@@ -39,7 +39,7 @@ if [ $SOURCED -ne 1 ]; then
 	unset SOURCEDIR
 	echo "Error: this script needs to be sourced in a supported shell" >&2
 	echo "Please check that the current shell is bash, zsh or ksh and run this script as '. $0 <args>'" >&2
-	return 1
+	exit 56
 else
 	unset SOURCED
 	tmpfile=$(mktemp /tmp/aglsetup.XXXXXXXX)
diff --git a/scripts/distro-manifest-generator.sh b/scripts/distro-manifest-generator.sh
index c7e8cc30..0db3e132 100755
--- a/scripts/distro-manifest-generator.sh
+++ b/scripts/distro-manifest-generator.sh
@@ -194,6 +194,9 @@ function _getgitmanifest() {
 	# Manifest build timestamp
 	DIST_BUILD_TS="$timestamp"
 
+	# build topic from setup topic
+	DIST_BUILD_TOPIC="${DIST_SETUP_TOPIC}"
+
 	# what to retain from setup manifest?
 	# to generate the full list: cat setup.manifest  | grep = | cut -f1 -d"=" | awk '{printf("%s ",$1);}'
 	declare -A SETUP_VARS
@@ -203,9 +206,9 @@ function _getgitmanifest() {
 
 	# extra vars not coming from setup.manifest but generated here
 	declare -A EXTRA_VARS
-	EXTRA_VARS[deploy]="DIST_SETUP_MANIFEST DIST_BUILD_TS DIST_LAYERS DIST_LAYERS_MD5 DIST_BUILD_HASH DIST_BUILD_ID"
-	EXTRA_VARS[target]="DIST_LAYERS DIST_BUILD_HASH DIST_BUILD_ID DIST_BUILD_TS"
-	EXTRA_VARS[sdk]="DIST_LAYERS DIST_BUILD_HASH DIST_BUILD_ID DIST_BUILD_TS"
+	EXTRA_VARS[deploy]="DIST_SETUP_MANIFEST DIST_BUILD_TS DIST_LAYERS DIST_LAYERS_MD5 DIST_BUILD_HASH DIST_BUILD_ID DIST_BUILD_TOPIC"
+	EXTRA_VARS[target]="DIST_LAYERS DIST_BUILD_HASH DIST_BUILD_ID DIST_BUILD_TS DIST_BUILD_TOPIC"
+	EXTRA_VARS[sdk]="DIST_LAYERS DIST_BUILD_HASH DIST_BUILD_ID DIST_BUILD_TS DIST_BUILD_TOPIC"
 
 	# BITBAKE_VARS may be defined from external file to source (--source arg)
 	# this is used to dump extra vars from inside bitbake recipe
-- 
2.20.1

